### PR TITLE
feat: remove the need for forRoot on material NgModules

### PR DESCRIPTION
**Pull request by jelbourn**
* Dialog and snackbar use their parent's state if there is one
* Other service that we want to be singletons and have state use a `useFactory` provider to get the ancestor and provide that if it exists

cc @IgorMinar can you provide thoughts on this?

cc FYI @josephperrott @mmalerba @andrewseguin @tinayuangao

### CODE REVIEW

**Comment by jelbourn** [comment]
Seeing some failures w/ AoT that don't quite make sense to me. Consulting w/ core team.

File: src/lib/button/button.ts [hunk 1]
```diff
@@ -167,10 +166,11 @@ export class MdAnchor extends MdButton {
   declarations: [MdButton, MdAnchor],
```
Comment (jelbourn) [inline src/lib/button/button.ts]: Nope, it is not included by way of `MdRippleModule`

Comment (kara) [inline src/lib/button/button.ts]: Cool, just double-checking

**Comment by jelbourn** [comment]
Also cc FYI @DevVersion @crisbeto @EladBezalel

**Comment by IgorMinar** [comment]
@jelbourn looks reasonable. please file an issue @ angular/angular if you haven't yet. Let's see if the core can help with this pattern as it looks better to me from the user perspective than the `forRoot` pattern.

**Comment by DevVersion** [comment]
@jelbourn Were the errors on the AOT mode valid? - Also nice change. Should avoid future provider issues as in #2538

**Comment by jelbourn** [comment]
@DevVersion yes, it caught some real errors

**Pull request merged**